Add code for 1D detector

ivan-usov opened the pull request.

JakHolzer committed:
Add files via upload

Try, except deleted.
Added one if/elif loop to tell apart zebra mode (bi, nb) and different angles saved

JakHolzer committed:
Add files via upload

function to find peaks (more detailed description in email)

JakHolzer committed:
Function to export *.comm files

JakHolzer committed:
peak fitting 

Fit one peak from ccl as combination of gaussian, lorentian and backgroud

ivan-usov committed:
Extract constants

ivan-usov committed:
Apply black formatter

ivan-usov committed:
Stop in case of malformed metadata section

Remove exception hadling because if there is an error in reading,
we shouldn't return partially read data, which will cause issues later.

JakHolzer committed:
Update load_1D.py

JakHolzer committed:
Update comm_export.py

JakHolzer committed:
Update fit2.py

JakHolzer committed:
guard close changed

JakHolzer committed:
removed strings, added argument

JakHolzer committed:
updated guard close, deleted plotting

JakHolzer committed:
Update ccl_findpeaks.py

ivan-usov committed:
Remove unused imports

JakHolzer committed:
Update ccl_findpeaks.py

JakHolzer committed:
Update fit2.py

corrected value range for numerical fit, added export_fit parameter, which should in future decide if exported value is fitted or integrated one.

JakHolzer committed:
Update fit2.py

While function to find different indexes if the peak is very thin and sigma is much smaller than step, added the find_nearest which was missing for some reason.

JakHolzer committed:
Update comm_export.py

reworked function to export data. Added lorenntz to decide if Lorentz correction should be done and also export_fit to decide whether fitter or itegrated area is to be exported (added also to fit2.py)

ivan-usov commented:
@JakHolzer , I created a pull request will all changes grouped on the same branch. This way it will be easier to introduce changes, that might affect all files at the same time. You should be able to push to that branch (`det1d`) too.

JakHolzer committed:
Update load_1D.py

Corrected the dat beaking bug

ivan-usov committed:
Revert "Update load_1D.py"

This reverts commit cc3f1d28a5670dad6fdaec83a19edbee0b1a09ce.

ivan-usov committed:
Break metadata and data reads

* fix decimal issue
* add unkwn_angle

ivan-usov committed:
Simplify dat file processing

* rename "omega" -> "om", and "counts" -> "Counts" for consistency
with the column names

ivan-usov commented on pyzebra/load_1D.py:
```diff
@@ -0,0 +1,172 @@
+        # read data
+        if det_variables["file_type"] == "ccl":
+            decimal = list()
+            data = infile.readlines()
```
@JakHolzer , if you want, you can try to simplify reading of ccl files in a similar way to https://github.com/paulscherrerinstitute/pyzebra/pull/11/commits/fe1d8044a3f2f6e24927a4f8baf4aa0894a56140

JakHolzer replied:
I will try 👍

ivan-usov replied:
Btw, that commit also renames "omega" into "om", and "counts" into "Counts", that is consistent with the column names in dat files. I hope it doesn't break you other code, otherwise let me know what names you prefer, it will be easy to fix, e.g. with
```
# row_names = next(infile).split()
row_names = ["NP", "omega", "counts", ...]
```

JakHolzer commented:
I think it should be fine. The ccl files does not have the headers and
there is only one data column, rest is numbers or is calculated, not read.
So I think its alright, important think is that you know what is what :)

út 15. 9. 2020 v 16:46 odesílatel Ivan Usov <notifications@github.com>
napsal:

> *@ivan-usov* commented on this pull request.
> ------------------------------
>
> In pyzebra/load_1D.py
> <https://github.com/paulscherrerinstitute/pyzebra/pull/11#discussion_r488727539>
> :
>
> > +                variable = variable.strip()
> +                if variable in META_VARS_FLOAT:
> +                    det_variables["meta"][variable] = float(value)
> +                elif variable in META_VARS_STR:
> +                    det_variables["meta"][variable] = str(value)[:-1].strip()
> +                elif variable in META_UB_MATRIX:
> +                    det_variables["meta"][variable] = re.findall(r"[-+]?\d*\.\d+|\d+", str(value))
> +
> +            if "#data" in line:
> +                # this is the end of metadata and the start of data section
> +                break
> +
> +        # read data
> +        if det_variables["file_type"] == "ccl":
> +            decimal = list()
> +            data = infile.readlines()
>
> Btw, that commit also renames "omega" into "om", and "counts" into
> "Counts", that is consistent with the column names in dat files. I hope it
> doesn't break you other code, otherwise let me know what names you prefer,
> it will be easy to fix, e.g. with
>
> # row_names = next(infile).split()
> row_names = ["NP", "omega", "counts", ...]
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/paulscherrerinstitute/pyzebra/pull/11#discussion_r488727539>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM2BBRVPO2FLEFKK74RXULLSF545RANCNFSM4RNBW7BQ>
> .
>

ivan-usov committed:
Use a ref to the current measurement

ivan-usov committed:
Check fit results instead of excepting KeyError

ivan-usov commented on pyzebra/comm_export.py:
```diff
@@ -0,0 +1,75 @@
+    align = ">"
+    if data["meta"]["indices"] == "hkl":
+        extension = ".comm"
+        padding = [6, 4, 10, 8]
```
These paddings look artificial to me. Do you think it was just tabs between numbers, and the different padding sizes are because of different lengths of those written numbers?

JakHolzer replied:
I'm sorry, I think that the answer I send from email got stuck somewhere. The files look very organized, so I think its really read line by line in this and first 6 chars correspond to number and so on. Oksana would like to keep it this way.

JakHolzer commented:
Well, the structure of the files seems very uniform and aligned, so I
wanted to keep that. I have never worked with the software that uses it, so
I'm not sure how it reads the data. My suggestion is to leave it like this
and we can ask Oksana to try load one file with padding and second with
tabs to see if it works.


st 16. 9. 2020 v 11:27 odesílatel Ivan Usov <notifications@github.com>
napsal:

> *@ivan-usov* commented on this pull request.
> ------------------------------
>
> In pyzebra/comm_export.py
> <https://github.com/paulscherrerinstitute/pyzebra/pull/11#discussion_r489296559>
> :
>
> > +            corr_value = np.abs(value * np.sin(ang1) * np.cos(ang2))
> +            return corr_value
> +
> +
> +def export_comm(data, path, lorentz=False):
> +    """exports data in the *.comm format
> +    :param lorentz: perform Lorentz correction
> +    :param path: path to file + name
> +    :arg data - data to export, is dict after peak fitting
> +
> +    """
> +
> +    align = ">"
> +    if data["meta"]["indices"] == "hkl":
> +        extension = ".comm"
> +        padding = [6, 4, 10, 8]
>
> These paddings look artificial to me. Do you think it was just tabs
> between numbers, and the different padding sizes are because of different
> lengths of those written numbers?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/paulscherrerinstitute/pyzebra/pull/11#pullrequestreview-489441678>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM2BBRU7NNDFJCK25MZNFQTSGCAJNANCNFSM4RNBW7BQ>
> .
>

ivan-usov committed:
Add parse_1D function

ivan-usov committed:
Import new functions on pyzebra import

ivan-usov committed:
Add panel placeholder for 1D detector processing

JakHolzer committed:
Update fit2.py

Added uncertainty propagation to the numerical integration and weighted fit. Case for 'no peak' scenario also added. Parameter export_fit in meas["fit"] added which is decides if fitted area or numerically integrated area will be saved in comm/incomm files.

JakHolzer commented on pyzebra/fit2.py:
```diff
@@ -0,0 +1,225 @@
+    print(result.fit_report())
+
+    print((result.params["g_amp"].value - int_area.n) / result.params["g_amp"].value)
+    d["export_fit"] = False
```
This parameter is (will be) used by export comm to decide if the fitted area or numerically integrated area is to be exported. User has to be able to choose, non the less numerically integrated area is default.

ivan-usov replied:
Does it need to be individual for each measurement? I would expect a user wanting to export all measurements in a similar way, thus having this parameter global.

JakHolzer replied:
Generally, user wants to export all fitted or integrated, but option to select individually in case of bad fit should be kept. However, I dont expect it to be used very often.

JakHolzer committed:
Update comm_export.py

Added the fit value/ num integrated value choice

JakHolzer committed:
Update load_1D.py

added the two unknown variables are temperature and field

JakHolzer committed:
Update fit2.py

ivan-usov committed:
Add a possiblity to select a measurement

ivan-usov committed:
Display basic peakfinding results

ivan-usov committed:
Add smooth option as an example

ivan-usov committed:
Display fitting results

ivan-usov committed:
Display content of exported file

ivan-usov committed:
Do not require matplotlib to be installed

JakHolzer committed:
Update fit2.py

fit needs at least 3 points to work properly, this should not fail again in case of extremely slim peaks

ivan-usov committed:
Simplify value references in export loop

ivan-usov committed:
Check for "fit" key instead of handling KeyError

JakHolzer committed:
functions to merge, scan, subtract measurements

JakHolzer committed:
Update ccl_dict_operation_v1.py

JakHolzer committed:
Visualize fit

Copy and paste into fit2 righ after:
  num_int_background = simps(background_errors, x[numfit_min:numfit_max])
  int_area = num_int_area-num_int_background

ca line 180

JakHolzer committed:
Add files via upload

Reuploaded from wrong folder

JakHolzer committed:
Delete ccl_dict_operation_v1.py

ivan-usov committed:
Add experimental code for saving data into a file

JakHolzer committed:
Update ccl_dict_operation.py

Corrected some lines, no major changes

ivan-usov committed:
Display gauss and bkg fits

ivan-usov committed:
Add input for proposal to simplify access data

JakHolzer committed:
Update ccl_dict_operation.py

added note to trace merging and sutraction

JakHolzer committed:
Update ccl_dict_operation.py

Added warning for negative numbers after subtraction

JakHolzer commented on pyzebra/load_1D.py:
```diff
@@ -0,0 +1,181 @@
+    "mcvl",
+    "momu",
+    "mcvu",
+    "detectorDistance",
```
the detectorDistance is a number in ccl files, but in dat files its date, which breaks the code. I asked Oksana if it can be fixed in the software that produces the *.dat files. If it for whatever reason cant be fixed, I suggest to move it to META_VARS_STR.

ivan-usov committed:
Rename panels

ivan-usov committed:
Replace measurement Select with DataTable

ivan-usov committed:
Add find peak controls

ivan-usov committed:
Split process into peak find and fit steps

ivan-usov committed:
Display number of peaks for measurements

JakHolzer committed:
Update comm_export.py

fixed some bugs

JakHolzer committed:
Update fit2.py

JakHolzer committed:
Update load_1D.py

added the data type to dict, it is necessary for adding dat files into ccl dicts

ivan-usov committed:
Add possibility to change peak position

ivan-usov committed:
Add peak fit controls

ivan-usov committed:
Add find peak and fit for currently selected data

ivan-usov committed:
Display fit results in concise way

ivan-usov committed:
Add fit param reset button

ivan-usov committed:
Display smoothed curve

ivan-usov committed:
Return preview of an export file

JakHolzer committed:
added more functions

JakHolzer committed:
Update fit2.py

Added check for stderr of fit area which can be None in case fit failure and can break the rest of the code.

ivan-usov committed:
Merge branch 'det1d' of https://github.com/paulscherrerinstitute/pyzebra into det1d

JakHolzer committed:
Update fit2.py

Added user input to no peak case

ivan-usov committed:
Fix imports and apply formatter

ivan-usov committed:
Parse peak_indexes as int

ivan-usov committed:
Add export_fit checkbox

ivan-usov committed:
Add integrate from .. to .. widgets

JakHolzer committed:
Update fit2.py

added option for data binning

ivan-usov committed:
Merge branch 'master' into det1d

ivan-usov committed:
Optimize layout

JakHolzer committed:
Modules for parametric study

Loading files from txt and list of filenames, creating and sorting the dataframe and making of the graphs

ivan-usov committed:
Display omega in hovertool

ivan-usov committed:
Display omega in hovertool

JakHolzer committed:
Update param_study_moduls.py

changed the delimiter to ","

JakHolzer committed:
Update fit2.py

fixed the weighted fit

JakHolzer committed:
Update fit2.py

added abs value to weights just in case

ivan-usov committed:
Cleanup, import sort, formatting

ivan-usov commented:
@JakHolzer , I would suggest not to extend this PR into parametric study, and merge it rather sooner than later. Do you have changes that are not related to parametric study still to be added here?

JakHolzer committed:
Update fit2.py

JakHolzer committed:
Update fit2.py

Second apology, that broke the code in another way, hopefully now its alrigh

ivan-usov committed:
Add basic display for stdout and bokeh logger

ivan-usov committed:
Replace export_fit checkbox with radiobutton group

The area method is now applying for all measurements

ivan-usov committed:
Compact hdf viewer panel

ivan-usov committed:
Use omega instead of a range sequence

ivan-usov committed:
Add vertical spans for integration borders

ivan-usov committed:
Fix manual peak_index entry

ivan-usov committed:
Refactor handling of the result dict

ivan-usov committed:
Rename "Measurements" to "meas"

ivan-usov committed:
Use integer keys for measurements

ivan-usov committed:
Allow fitting of measurements without peaks

ivan-usov committed:
Fix integ values not applying in Fit Current

ivan-usov committed:
Fix issue reading ccl files with 5 digit counts

JakHolzer committed:
Update load_1D.py

changed the dat part

ivan-usov committed:
Refactor parsing of the first lines in each meas

ivan-usov commented on pyzebra/load_1D.py:
```diff
@@ -165,7 +165,15 @@ def parse_1D(fileobj, data_type):
+        data_cols['omega_angle'] = metadata['omega']
+        data_cols['number_of_measurements'] = len(data_cols['om'])
+        data_cols['monitor'] = data_cols['Monitor1'][0]
+        measurements["1"] = dict(data_cols)
```
The key should be just an integer 1, e.g. `measurements[1] = dict(data_cols)`

ivan-usov committed:
Refactor read of counts and clarify d["om"]

ivan-usov commented on pyzebra/load_1D.py:
```diff
@@ -0,0 +1,212 @@
+            for param, (param_name, param_type) in zip(params, ccl_second_line):
+                d[param_name] = param_type(param)
+
+            d["om"] = np.linspace(
```
@JakHolzer , could you confirm that this is correct, especially when the number of measurements is odd

JakHolzer replied:
I believe so, I've tested that now on some model data and it seems everything is in order. The omega angle in the first line (not the array) is centre of the the scan and doesent have to be in the omega array. Ive tried it on odd and even numbers and mean of omega array is equal to omega value. 
for pints in 20, 21:
    omega  = 31.4
    step = 0.1

    om = np.linspace(float(omega) - (int(pints) / 2) * float(step),
        float(omega) + (int(pints) / 2) * float(step),
        int(pints))

    print(om)
    print(np.mean(om))

gives:
[30.4        30.50526316 30.61052632 30.71578947 30.82105263 30.92631579
 31.03157895 31.13684211 31.24210526 31.34736842 31.45263158 31.55789474
 31.66315789 31.76842105 31.87368421 31.97894737 32.08421053 32.18947368
 32.29473684 32.4       ]
31.4
[30.35  30.455 30.56  30.665 30.77  30.875 30.98  31.085 31.19  31.295
 31.4   31.505 31.61  31.715 31.82  31.925 32.03  32.135 32.24  32.345
 32.45 ]
31.4

In odd, the omega is middle value, in even case the omega itself is not there. Or am I missing something?

ivan-usov replied:
OK, I didn't know how this "om" array should be constructed at all. This is clearly dependent on how daq works, so if you know that this array is calculated this way, then it's all good.

JakHolzer commented:
Oh I see, I apologise. Thank you for so throughout correction :)

po 19. 10. 2020 v 15:44 odesílatel Ivan Usov <notifications@github.com>
napsal:

> *@ivan-usov* commented on this pull request.
> ------------------------------
>
> In pyzebra/load_1D.py
> <https://github.com/paulscherrerinstitute/pyzebra/pull/11#discussion_r507738757>
> :
>
> > @@ -165,7 +165,15 @@ def parse_1D(fileobj, data_type):
>              for name, val in zip(col_names, line.split()):
>                  data_cols[name].append(float(val))
>
> -        measurements = dict(data_cols)
> +        data_cols['h_index'] = float(metadata['title'].split()[-3])
> +        data_cols['k_index'] = float(metadata['title'].split()[-2])
> +        data_cols['l_index'] = float(metadata['title'].split()[-1])
> +        data_cols['temperature'] = metadata['temp']
> +        data_cols['mag_field'] = metadata['mf']
> +        data_cols['omega_angle'] = metadata['omega']
> +        data_cols['number_of_measurements'] = len(data_cols['om'])
> +        data_cols['monitor'] = data_cols['Monitor1'][0]
> +        measurements["1"] = dict(data_cols)
>
> The key should be just an integer 1, e.g. measurements[1] =
> dict(data_cols)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/paulscherrerinstitute/pyzebra/pull/11#pullrequestreview-511751197>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM2BBRUQGCFAYYI3DQJYJ2DSLQ7CDANCNFSM4RNBW7BQ>
> .
>

JakHolzer commented:
Well, its not necessary for dat files which actually carry the om array, in
ccl its simplified to omega centre, step and number of points, so this is
only way of getting it.

po 19. 10. 2020 v 16:12 odesílatel Ivan Usov <notifications@github.com>
napsal:

> *@ivan-usov* commented on this pull request.
> ------------------------------
>
> In pyzebra/load_1D.py
> <https://github.com/paulscherrerinstitute/pyzebra/pull/11#discussion_r507781946>
> :
>
> > +            # first line
> +            measurement_number, *params = line.split()
> +            for param, (param_name, param_type) in zip(params, ccl_first_line):
> +                d[param_name] = param_type(param)
> +
> +            decimal.append(bool(Decimal(d["h_index"]) % 1 == 0))
> +            decimal.append(bool(Decimal(d["k_index"]) % 1 == 0))
> +            decimal.append(bool(Decimal(d["l_index"]) % 1 == 0))
> +
> +            # second line
> +            next_line = next(fileobj)
> +            params = next_line.split()
> +            for param, (param_name, param_type) in zip(params, ccl_second_line):
> +                d[param_name] = param_type(param)
> +
> +            d["om"] = np.linspace(
>
> OK, I didn't know how this "om" array should be constructed at all. This
> is clearly dependent on how daq works, so if you know that this array is
> calculated this way, then it's all good.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/paulscherrerinstitute/pyzebra/pull/11#discussion_r507781946>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM2BBRTHNOPYSCPDKIF76KDSLRCMZANCNFSM4RNBW7BQ>
> .
>

ivan-usov committed:
Add debug script

JakHolzer committed:
Update fit2.py

moved the check up so it catches it  right after fitting. I think this should solve the problem.

ivan-usov committed:
Don't stop on measurements with failed fit

Fix #13

ivan-usov committed:
Add fit results to the table

ivan-usov committed:
Revert "Display omega in hovertool"

This reverts commit 48efa93234cd76553a11a2441e27c676ae915826.

JakHolzer committed:
Update param_study_moduls.py

Should work with new dict keys

ivan-usov committed:
Fix image not updating

JakHolzer committed:
Update load_1D.py

Replaced the "1" to 1 in dat part

JakHolzer committed:
Update load_1D.py

Added angles from meta to each scan, try except to try to get hkl from title

JakHolzer committed:
Update load_1D.py

Correcting my stupidity

ivan-usov committed:
Revert "Update load_1D.py"

This reverts commit f79553c1ba4135acbd7f758d5071e57757b31d98.

ivan-usov committed:
Reapply changes

ivan-usov committed:
Display std dev for counts

ivan-usov committed:
Make found peaks more visible

ivan-usov committed:
Add hkl indices to scan table

ivan-usov committed:
Use 'scan' instead of 'meas' for array of counts
